Applications/FileIO/SWMM/SWMMInterface.cpp

```diff
@@ -0,0 +1,1248 @@
+        if (o_it == point_names.end())
+        {
+            ERR ("SwmmInterface::readLineElements(): Outlet node %s not found coordinates map.", outlet.c_str());
+            return false;
+        }
+        GeoLib::Polyline* ply = new GeoLib::Polyline(points);
+        std::size_t a (i_it->second);
+        ply->addPoint(i_it->second);
+        ply->addPoint(o_it->second);
+        lines.push_back(ply);
+        line_names.push_back(split_str[0]);
+        getline(in, line);
+    }
+    return true;
+}
+
+/// Deletes the geometric objects and returns false
+bool geometryCleanup(std::vector<GeoLib::Point*> &points, std::vector<GeoLib::Polyline*> &lines)
+{
+    for (auto line : lines)
+        delete line;
+    for (auto point : points)
+        delete point;
+    return false;
+}
+
+bool SwmmInterface::SwmmInputToGeometry(std::string const& inp_file_name, GeoLib::GEOObjects &geo_objects, bool add_subcatchments)
+{
+    if (!isSwmmInputFile(inp_file_name))
+        return false;
+
+    std::ifstream in ( inp_file_name.c_str() );
+    std::unique_ptr<std::vector<GeoLib::Point*>> points (new std::vector<GeoLib::Point*>);
+    std::unique_ptr<std::vector<GeoLib::Polyline*>> lines (new std::vector<GeoLib::Polyline*>);
+    std::vector<std::string> pnt_names;
+    std::vector<std::string> line_names;
+
+    std::string geo_name = BaseLib::extractBaseNameWithoutExtension(inp_file_name);
+    std::string line;
+    while ( getline(in, line) )
+    {
+        if (line == "[COORDINATES]")
+        {
+            if (!readCoordinates<GeoLib::Point>(in, *points, pnt_names))
+                return geometryCleanup(*points, *lines);
+        }
+        if (line == "[VERTICES]")
+        {
+            if (!readCoordinates<GeoLib::Point>(in, *points, pnt_names))
+                return geometryCleanup(*points, *lines);
+        }
+        if (line == "[Polygons]" && add_subcatchments)
+        {
+            if (!readPolygons(in, *lines, line_names, *points, pnt_names))
+                return geometryCleanup(*points, *lines);
+        }
+        if (line == "[SYMBOLS]")
+        {
+            if (!readCoordinates<GeoLib::Point>(in, *points, pnt_names))
+                return geometryCleanup(*points, *lines);
+        }
+    }
+
+    if (points->empty())
+    {
+        ERR ("No points found in file");
+        return false;
+    }
+    if (points->size() != pnt_names.size())
+    {
+        ERR ("Lengt of point vector and point name vector do not match.");
+        return geometryCleanup(*points, *lines);
+    }
+
+    std::map<std::string, std::size_t> *name_id_map (new std::map<std::string, std::size_t>);
+    std::size_t const n_names (pnt_names.size());
+    for (std::size_t i=0; i<n_names; ++i)
+    {
+        if (pnt_names[i] != "")
+            name_id_map->insert(std::make_pair(pnt_names[i], i));
+    }
+
+    // rewind stream and read links between junctions
+    in.clear();
+    in.seekg(0, in.beg);
+
+    while ( getline(in, line) )
```

Comment:
This is a good pattern for reading streams (because of the checked stream state). Nice!